File: nodejs/tests/motor/drv8830_test_unit.js
'use strict';
const { I2CConnectionMock } = require('../../packages/periph/src/connection/i2c_mock');
const { DRV8830Minimal, DRV8830Full } = require('../../packages/periph/src/chips/motor/drv8830');

const _REG_CONTROL = 0x00;
const _REG_FAULT   = 0x01;

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else { console.log('FAIL', label); failed++; }
}

function checkEq(label, got, expected) {
    if (got === expected) { console.log('PASS', label); passed++; }
    else { console.log(`FAIL ${label}: got ${got}, expected ${expected}`); failed++; }
}

function lastWrite(conn, reg) {
    const w = conn.writes.filter((b) => b.length === 2 && b[0] === reg);
    return w.length ? w[w.length - 1][1] : undefined;
}

async function main() {
    // -- init(): plain presence read, no register writes -----------------
    {
        const conn = new I2CConnectionMock();
        const motor = new DRV8830Minimal(conn);
        await motor.init();
        checkEq('init_probe_is_address_byte_only', conn.writes.length, 1);
        checkEq('init_probe_reads_control', conn.writes[0][0], _REG_CONTROL);
    }

    // -- drive(): conversion, direction bits, floor and clamp ------------
    {
        const conn = new I2CConnectionMock();
        const motor = new DRV8830Minimal(conn);
        await motor.drive(3.0);
        checkEq('drive_forward_3v', lastWrite(conn, _REG_CONTROL), (37 << 2) | 0x01);
        await motor.drive(-2.0);
        checkEq('drive_reverse_2v', lastWrite(conn, _REG_CONTROL), (25 << 2) | 0x02);
        await motor.drive(0);
        checkEq('drive_zero_coasts', lastWrite(conn, _REG_CONTROL), 0x00);
        await motor.drive(0.4);
        checkEq('drive_below_floor_coasts', lastWrite(conn, _REG_CONTROL), 0x00);
        await motor.drive(0.48);
        checkEq('drive_floor_vset6', lastWrite(conn, _REG_CONTROL), (6 << 2) | 0x01);
        await motor.drive(9.0);
        checkEq('drive_clamps_to_vset63', lastWrite(conn, _REG_CONTROL), (63 << 2) | 0x01);
        await motor.brake();
        checkEq('brake_writes_0x03', lastWrite(conn, _REG_CONTROL), 0x03);
        await motor.stop();
        checkEq('stop_writes_0x00', lastWrite(conn, _REG_CONTROL), 0x00);
        checkTrue('minimal_has_no_read_fault', typeof motor.readFault === 'undefined');
    }

    // -- setOutput(): raw fields, reserved codes rejected ----------------
    {
        const conn = new I2CConnectionMock();
        const motor = new DRV8830Full(conn);
        await motor.setOutput(20, false, true);
        checkEq('set_output_raw', lastWrite(conn, _REG_CONTROL), (20 << 2) | 0x02);
        let threw = false;
        try { await motor.setOutput(5, true, false); } catch (e) { threw = e instanceof RangeError; }
        checkTrue('set_output_rejects_reserved', threw);
    }

    // -- readOutput(): decode VSET and direction -------------------------
    {
        const conn = new I2CConnectionMock();
        const motor = new DRV8830Full(conn);
        conn.setRegister(_REG_CONTROL, [(63 << 2) | 0x01]);
        let out = await motor.readOutput();
        checkTrue('read_output_forward', out.direction === 'forward' && Math.abs(out.voltage - 5.06) < 0.01);
        conn.setRegister(_REG_CONTROL, [(16 << 2) | 0x02]);
        out = await motor.readOutput();
        checkTrue('read_output_reverse', out.direction === 'reverse' && Math.abs(out.voltage - 1.285) < 0.001);
        conn.setRegister(_REG_CONTROL, [0x03]);
        out = await motor.readOutput();
        checkTrue('read_output_brake', out.direction === 'brake' && out.voltage === 0);
        conn.setRegister(_REG_CONTROL, [0x00]);
        out = await motor.readOutput();
        checkTrue('read_output_coast', out.direction === 'coast' && out.voltage === 0);
    }

    // -- readFault()/pollInterrupt()/clearFault() ------------------------
    {
        const conn = new I2CConnectionMock();
        conn.setRegister(_REG_FAULT, [0x01 | 0x10]);
        const motor = new DRV8830Full(conn);
        let f = await motor.readFault();
        checkTrue('read_fault_ilimit', f.fault && !f.ocp && !f.uvlo && !f.ots && f.ilimit);
        conn.setRegister(_REG_FAULT, [0x01 | 0x02 | 0x04 | 0x08]);
        f = await motor.pollInterrupt();
        checkTrue('poll_interrupt_ocp_uvlo_ots', f.fault && f.ocp && f.uvlo && f.ots && !f.ilimit);
        checkEq('read_fault_does_not_clear', lastWrite(conn, _REG_FAULT), undefined);
        await motor.clearFault();
        checkEq('clear_fault_writes_0x80', lastWrite(conn, _REG_FAULT), 0x80);
    }

    // -- onInterrupt(): polling fallback fires once per new fault --------
    {
        const conn = new I2CConnectionMock();
        const motor = new DRV8830Full(conn);
        const calls = [];
        await motor.onInterrupt((s) => calls.push(s));
        await new Promise((r) => setTimeout(r, 30));
        conn.setRegister(_REG_FAULT, [0x01 | 0x08]);
        await new Promise((r) => setTimeout(r, 40));
        await motor.offInterrupt();
        checkEq('on_interrupt_fires_once_per_fault', calls.length, 1);
        checkTrue('on_interrupt_passes_fault_status', calls.length === 1 && calls[0].ots);
    }

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();

File: python/tests/motor/drv8830_test_unit.py
"""Unit test for the DRV8830 — runs without hardware using the I2C mock.

Verifies voltage-to-VSET conversion, direction bit encoding, the coast
floor and clamp, raw CONTROL writes, output read-back decoding, and fault
reporting/clearing by observing what the driver writes to the mock.
"""

from periph.connection.i2c_mock import I2CConnectionMock
from periph.chips.motor.drv8830 import DRV8830Minimal, DRV8830Full

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


def last_write(mock, reg):
    return [w[1] for w in mock.writes if len(w) == 2 and w[0] == reg][-1]


REG_CONTROL = DRV8830Minimal._REG_CONTROL
REG_FAULT = DRV8830Minimal._REG_FAULT

# Construction makes no writes (presence check is a read only).
mock = I2CConnectionMock()
motor = DRV8830Minimal(mock)
check_true('init_no_register_writes', all(len(w) == 1 for w in mock.writes))

# drive(3.0): VSET = round(3.0 * 16 / 1.285) = 37 (0x25), forward (IN1).
motor.drive(3.0)
check_true('drive_forward_3v', last_write(mock, REG_CONTROL) == (37 << 2) | 0x01)

# drive(-2.0): VSET = round(2.0 * 16 / 1.285) = 25 (0x19), reverse (IN2).
motor.drive(-2.0)
check_true('drive_reverse_2v', last_write(mock, REG_CONTROL) == (25 << 2) | 0x02)

# drive(0.0) and sub-floor magnitudes coast with CONTROL = 0x00.
motor.drive(0.0)
check_true('drive_zero_coasts', last_write(mock, REG_CONTROL) == 0x00)
motor.drive(0.4)
check_true('drive_below_floor_coasts', last_write(mock, REG_CONTROL) == 0x00)

# drive(0.48): VSET 6 — the minimum valid code.
motor.drive(0.48)
check_true('drive_floor_vset6', last_write(mock, REG_CONTROL) == (6 << 2) | 0x01)

# drive(9.0): clamped to VSET 63.
motor.drive(9.0)
check_true('drive_clamps_to_vset63', last_write(mock, REG_CONTROL) == (63 << 2) | 0x01)

motor.brake()
check_true('brake_writes_0x03', last_write(mock, REG_CONTROL) == 0x03)
motor.stop()
check_true('stop_writes_0x00', last_write(mock, REG_CONTROL) == 0x00)

# Minimal does not expose the Full API.
check_true('minimal_has_no_read_fault', not hasattr(motor, 'read_fault'))

# set_output: raw fields; reserved codes rejected.
mock2 = I2CConnectionMock()
full = DRV8830Full(mock2)
full.set_output(20, False, True)
check_true('set_output_raw', last_write(mock2, REG_CONTROL) == (20 << 2) | 0x02)
try:
    full.set_output(5, True, False)
    check_true('set_output_rejects_reserved', False)
except ValueError:
    check_true('set_output_rejects_reserved', True)

# read_output decodes VSET and direction.
mock3 = I2CConnectionMock()
mock3.set_register(REG_CONTROL, (63 << 2) | 0x01)
full3 = DRV8830Full(mock3)
v, d = full3.read_output()
check_true('read_output_forward', d == 'forward' and abs(v - 5.06) < 0.01)
mock3.set_register(REG_CONTROL, (16 << 2) | 0x02)
v, d = full3.read_output()
check_true('read_output_reverse', d == 'reverse' and abs(v - 1.285) < 0.001)
mock3.set_register(REG_CONTROL, 0x03)
check_true('read_output_brake', full3.read_output() == (0.0, 'brake'))
mock3.set_register(REG_CONTROL, 0x00)
check_true('read_output_coast', full3.read_output() == (0.0, 'coast'))

# read_fault decodes each bit; does not write (never clears implicitly).
mock4 = I2CConnectionMock()
mock4.set_register(REG_FAULT, 0x01 | 0x10)  # FAULT + ILIMIT
full4 = DRV8830Full(mock4)
check_true('read_fault_ilimit', full4.read_fault() == (True, False, False, False, True))
mock4.set_register(REG_FAULT, 0x01 | 0x02 | 0x04 | 0x08)
check_true('poll_interrupt_ocp_uvlo_ots', full4.poll_interrupt() == (True, True, True, True, False))
check_true('read_fault_does_not_clear', not any(len(w) == 2 and w[0] == REG_FAULT for w in mock4.writes))

full4.clear_fault()
check_true('clear_fault_writes_0x80', last_write(mock4, REG_FAULT) == 0x80)

print('===DONE: %d passed, %d failed===' % (passed, failed))
if failed:
    raise SystemExit(1)
